#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
gRPC server for fortune-analysis-service.
é¢ç›¸æ‰‹ç›¸å‘½ç†åˆ†ææœåŠ¡
"""

from __future__ import annotations

import os
import sys
from concurrent import futures

import grpc

BASE_DIR = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
PROJECT_ROOT = BASE_DIR
sys.path.insert(0, PROJECT_ROOT)

# å¯¼å…¥ç”Ÿæˆçš„ gRPC ä»£ç 
sys.path.insert(0, os.path.join(PROJECT_ROOT, "proto", "generated"))
import fortune_analysis_pb2
import fortune_analysis_pb2_grpc

from services.fortune_analysis.hand_analyzer_core import HandAnalyzerCore
from services.fortune_analysis.face_analyzer_core import FaceAnalyzerCore


class FortuneAnalysisServicer(fortune_analysis_pb2_grpc.FortuneAnalysisServiceServicer):
    """å®ç° FortuneAnalysisService çš„ gRPC æœåŠ¡"""

    def __init__(self):
        # ä½¿ç”¨ç‹¬ç«‹çš„æ‰‹ç›¸å’Œé¢ç›¸åˆ†æå™¨ï¼Œäº’ä¸å½±å“
        self.hand_analyzer = HandAnalyzerCore()
        self.face_analyzer = FaceAnalyzerCore()

    def AnalyzeHand(self, request: fortune_analysis_pb2.HandAnalysisRequest, context: grpc.ServicerContext) -> fortune_analysis_pb2.HandAnalysisResponse:
        """æ‰‹ç›¸åˆ†æ"""
        import datetime
        request_time = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        print(f"[{request_time}] ğŸ“¥ fortune-analysis-service: æ”¶åˆ°æ‰‹ç›¸åˆ†æè¯·æ±‚", flush=True)
        
        try:
            # è°ƒç”¨æ‰‹ç›¸åˆ†æå™¨ï¼ˆç‹¬ç«‹æ¨¡å—ï¼‰
            result = self.hand_analyzer.analyze_hand(
                image_bytes=request.image.image_bytes,
                image_format=request.image.image_format or "jpg",
                bazi_info=request.bazi_info if request.bazi_info.use_bazi else None
            )
            
            # æ„å»ºå“åº”
            response = fortune_analysis_pb2.HandAnalysisResponse()
            response.success = result.get("success", False)
            
            if not response.success:
                response.error_message = result.get("error", "åˆ†æå¤±è´¥")
                return response
            
            # å¡«å……ç‰¹å¾
            features = result.get("features", {})
            response.features.hand_shape = features.get("hand_shape", "")
            if features.get("finger_lengths"):
                response.features.finger_lengths.update(features["finger_lengths"])
            if features.get("palm_lines"):
                response.features.palm_lines.update(features["palm_lines"])
            if features.get("measurements"):
                response.features.measurements.update(features["measurements"])
            
            # å¡«å……æ´å¯Ÿ
            for insight in result.get("insights", []):
                insight_pb = fortune_analysis_pb2.AnalysisInsight(
                    category=insight.get("category", ""),
                    content=insight.get("content", ""),
                    confidence=insight.get("confidence", 0.0),
                    source=insight.get("source", "hand")
                )
                response.insights.append(insight_pb)
            
            # å¡«å……èåˆæ´å¯Ÿ
            for insight in result.get("integrated_insights", []):
                insight_pb = fortune_analysis_pb2.AnalysisInsight(
                    category=insight.get("category", ""),
                    content=insight.get("content", ""),
                    confidence=insight.get("confidence", 0.0),
                    source=insight.get("source", "integrated")
                )
                response.integrated_insights.append(insight_pb)
            
            # å¡«å……å»ºè®®
            response.recommendations.extend(result.get("recommendations", []))
            response.confidence = result.get("confidence", 0.0)
            
            # å®Œæ•´æŠ¥å‘Šï¼ˆJSONï¼‰
            import json
            response.report_json = json.dumps(result, ensure_ascii=False)
            
            print(f"[{request_time}] âœ… fortune-analysis-service: æ‰‹ç›¸åˆ†æå®Œæˆ", flush=True)
            return response
            
        except Exception as e:
            import traceback
            error_msg = f"æ‰‹ç›¸åˆ†æå¤±è´¥: {str(e)}\n{traceback.format_exc()}"
            print(f"[{request_time}] âŒ fortune-analysis-service: é”™è¯¯ - {error_msg}", flush=True)
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"æ‰‹ç›¸åˆ†æå¤±è´¥: {str(e)}")
            response = fortune_analysis_pb2.HandAnalysisResponse()
            response.success = False
            response.error_message = str(e)
            return response

    def AnalyzeFace(self, request: fortune_analysis_pb2.FaceAnalysisRequest, context: grpc.ServicerContext) -> fortune_analysis_pb2.FaceAnalysisResponse:
        """é¢ç›¸åˆ†æ"""
        import datetime
        request_time = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        print(f"[{request_time}] ğŸ“¥ fortune-analysis-service: æ”¶åˆ°é¢ç›¸åˆ†æè¯·æ±‚", flush=True)
        
        try:
            # è°ƒç”¨é¢ç›¸åˆ†æå™¨ï¼ˆç‹¬ç«‹æ¨¡å—ï¼‰
            result = self.face_analyzer.analyze_face(
                image_bytes=request.image.image_bytes,
                image_format=request.image.image_format or "jpg",
                bazi_info=request.bazi_info if request.bazi_info.use_bazi else None
            )
            
            # æ„å»ºå“åº”
            response = fortune_analysis_pb2.FaceAnalysisResponse()
            response.success = result.get("success", False)
            
            if not response.success:
                response.error_message = result.get("error", "åˆ†æå¤±è´¥")
                return response
            
            # å¡«å……ç‰¹å¾
            features = result.get("features", {})
            if features.get("san_ting_ratio"):
                response.features.san_ting_ratio.update(features["san_ting_ratio"])
            if features.get("facial_attributes"):
                response.features.facial_attributes.update(features["facial_attributes"])
            if features.get("feature_measurements"):
                response.features.feature_measurements.update(features["feature_measurements"])
            response.features.special_features.extend(features.get("special_features", []))
            
            # å¡«å……æ´å¯Ÿ
            for insight in result.get("insights", []):
                insight_pb = fortune_analysis_pb2.AnalysisInsight(
                    category=insight.get("category", ""),
                    content=insight.get("content", ""),
                    confidence=insight.get("confidence", 0.0),
                    source=insight.get("source", "face")
                )
                response.insights.append(insight_pb)
            
            # å¡«å……èåˆæ´å¯Ÿ
            for insight in result.get("integrated_insights", []):
                insight_pb = fortune_analysis_pb2.AnalysisInsight(
                    category=insight.get("category", ""),
                    content=insight.get("content", ""),
                    confidence=insight.get("confidence", 0.0),
                    source=insight.get("source", "integrated")
                )
                response.integrated_insights.append(insight_pb)
            
            # å¡«å……å»ºè®®
            response.recommendations.extend(result.get("recommendations", []))
            response.confidence = result.get("confidence", 0.0)
            
            # å®Œæ•´æŠ¥å‘Šï¼ˆJSONï¼‰
            import json
            response.report_json = json.dumps(result, ensure_ascii=False)
            
            print(f"[{request_time}] âœ… fortune-analysis-service: é¢ç›¸åˆ†æå®Œæˆ", flush=True)
            return response
            
        except Exception as e:
            import traceback
            error_msg = f"é¢ç›¸åˆ†æå¤±è´¥: {str(e)}\n{traceback.format_exc()}"
            print(f"[{request_time}] âŒ fortune-analysis-service: é”™è¯¯ - {error_msg}", flush=True)
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"é¢ç›¸åˆ†æå¤±è´¥: {str(e)}")
            response = fortune_analysis_pb2.FaceAnalysisResponse()
            response.success = False
            response.error_message = str(e)
            return response

    def HealthCheck(self, request: fortune_analysis_pb2.HealthCheckRequest, context: grpc.ServicerContext) -> fortune_analysis_pb2.HealthCheckResponse:
        """å¥åº·æ£€æŸ¥"""
        return fortune_analysis_pb2.HealthCheckResponse(status="ok")


def serve(port: int = 9005):
    """å¯åŠ¨ gRPC æœåŠ¡å™¨ï¼ˆæ”¯æŒçƒ­æ›´æ–°ï¼‰"""
    try:
        from server.hot_reload.microservice_reloader import (
            create_hot_reload_server,
            register_microservice_reloader
        )
        
        server_options = [
            ('grpc.keepalive_time_ms', 300000),
            ('grpc.keepalive_timeout_ms', 20000),
            ('grpc.keepalive_permit_without_calls', False),
            ('grpc.http2.max_pings_without_data', 2),
            ('grpc.http2.min_time_between_pings_ms', 60000),
        ]
        
        server, reloader = create_hot_reload_server(
            service_name="fortune_analysis",
            module_path="services.fortune_analysis.grpc_server",
            servicer_class_name="FortuneAnalysisServicer",
            add_servicer_to_server_func=fortune_analysis_pb2_grpc.add_FortuneAnalysisServiceServicer_to_server,
            port=port,
            server_options=server_options,
            max_workers=10,
            check_interval=30
        )
        
        register_microservice_reloader("fortune_analysis", reloader)
        reloader.start()
        
        # create_hot_reload_server å·²ç»ç»‘å®šäº†ç«¯å£ï¼Œä¸éœ€è¦å†æ¬¡ç»‘å®š
        server.start()
        print(f"âœ… Fortune Analysis gRPC æœåŠ¡å·²å¯åŠ¨ï¼ˆçƒ­æ›´æ–°å·²å¯ç”¨ï¼‰ï¼Œç›‘å¬ç«¯å£: {port}")
        
        try:
            server.wait_for_termination()
        except KeyboardInterrupt:
            print("\n>>> æ­£åœ¨åœæ­¢æœåŠ¡...")
            reloader.stop()
            server.stop(grace=5)
            print("âœ… æœåŠ¡å·²åœæ­¢")
            
    except ImportError:
        # é™çº§åˆ°ä¼ ç»Ÿæ¨¡å¼
        server_options = [
            ('grpc.keepalive_time_ms', 300000),
            ('grpc.keepalive_timeout_ms', 20000),
            ('grpc.keepalive_permit_without_calls', False),
            ('grpc.http2.max_pings_without_data', 2),
            ('grpc.http2.min_time_between_pings_ms', 60000),
        ]
        
        server = grpc.server(
            futures.ThreadPoolExecutor(max_workers=10),
            options=server_options
        )
        fortune_analysis_pb2_grpc.add_FortuneAnalysisServiceServicer_to_server(FortuneAnalysisServicer(), server)
        
        listen_addr = f"[::]:{port}"
        server.add_insecure_port(listen_addr)
        
        server.start()
        print(f"âœ… Fortune Analysis gRPC æœåŠ¡å·²å¯åŠ¨ï¼ˆä¼ ç»Ÿæ¨¡å¼ï¼‰ï¼Œç›‘å¬ç«¯å£: {port}")
        
        try:
            server.wait_for_termination()
        except KeyboardInterrupt:
            print("\n>>> æ­£åœ¨åœæ­¢æœåŠ¡...")
            server.stop(grace=5)
            print("âœ… æœåŠ¡å·²åœæ­¢")


if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser(description="å¯åŠ¨ Fortune Analysis gRPC æœåŠ¡")
    parser.add_argument("--port", type=int, default=9005, help="æœåŠ¡ç«¯å£ï¼ˆé»˜è®¤: 9005ï¼‰")
    args = parser.parse_args()
    serve(args.port)

